dpgen/generator/model_devi.py

```diff
@@ -0,0 +1,191 @@
+        try:
+            return ModelDeviEngien.engiens()[key]
+        except IndexError as e:
+            raise RuntimeError("Unsupported engien!") from e
+
+    def __init__(self, jdata: dict, mdata: dict):
+        self.jdata = jdata
+        self.mdata = mdata
+
+    @abstractmethod
+    def make_input(self, iter_idx:int, sys_idx: int, directory: Iterator[str], conf_name: str, models: List[str]):
+        """Make a simulation input from a directory.
+
+        The length of system should be only 1.
+        
+        Parameters
+        ----------
+        iter_idx: int
+            index of iteration
+        sys_idx: int
+            index of system
+        directory: Iterator
+            generate the working directory to run simulations
+        conf_name: str
+            file name of initial system
+        models: list[str]
+            list of models
+        """
+        raise NotImplementedError("Not implemented")
+
+    @abstractmethod
+    def get_running_parameters(self, work_path: str) -> Tuple[str, List[str]]:
+        """Get running parameters.
+
+        Parameters
+        ----------
+        work_path: str
+            the work path
+
+        Returns
+        -------
+        command: str
+            the running command
+        forward_files: list of str
+            forward files
+        backward_files: list of str
+            backward files
+        common_files: list of str
+            common files
+        """
+        raise NotImplementedError("Not implemented")
+
+
+    @abstractmethod
+    def extract_trajectory(self, directory) -> 'Trajectory':
+        """Obtain a Trajectory object from a directory.
+        
+        Returns
+        -------
+        trajectory: Trajectory
+            the Trajectory object
+        """
+        return Trajectory(directory)
+
+
+class Trajectory(ABC):
+    """A trajectory containing lots of frames or clusters.
+    
+    Parameters
+    ----------
+    engien: ModelDeviEngien
+        Engien
+    directory: str
+        The directory of a simulation
+    """
+    def __init__(self, engien: ModelDeviEngien, directory: str) -> None:
+        self.engien = engien
+        self.directory = directory
+        self.status = 0
+    
+    @abstractmethod
+    def get_model_deviations(self) -> np.ndarray:
+        """Get model deviations of all frames or clusters.
+
+        Returns
+        -------
+        model_deviations: np.ndarray[float], 1D or 2D
+            Model deviation of all frames or all clusters. 
+            - frame: 1D, the maximum model deviation of all atoms
+            - clusters: 2D, the atomic model deviation of centeral atom
+        
+        Notes
+        -----
+        The index of model deviation should be as the same as frames.
+        """
+        # for performance, returns a numpy array instead of a list
+        raise NotImplementedError("Not implemented")
+    
+    @abstractmethod
+    def get_frame(self, idx: Tuple[int]) -> "Frame":
```

Comment:
get single frame. so `idx:int` be `tuple` ?

Reply:
Fixed.
A new `Trajectory` should implement thier own `Frame`, as the current `Frame` do nothing in `read_frame` method. So this should be an abstractmethod.